### PR TITLE
[C++ API] fix wrong arg handling for object references in `TaskExecutor`, `TaskArgByReference`

### PR DESCRIPTION
**Pull request by jon-chuang**
## Why are these changes needed?
Previously, ref arg is handled wrongly, serializing the object ref, instead of RayObject to be passed as args buffer to the user function. 

That's because CoreWorker is the component responsible for ensuring that all ObjectReferences are resolved and serialized into `RayObject`s at the time of the `task_execution_callback` invocation, not any component downstream of the callback. 

This resulted in the following error for large objects which are not turned into `TaskArg::value` due to being over 100KB.
```
C++ exception with description "Invalid: invalid arguments: std::bad_cast" thrown in the test body.
```
This was not caught due to lack of testing for large objects, which has now been added.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jon-chuang** [comment]
CI reports a SIGSEGV, investigating...

**Comment by jon-chuang** [comment]
I'm running into additional spurious errors even with this fix, leading me to think there is something amiss with the C++ implementation for large objects

**Comment by jon-chuang** [comment]
@SongGuyang , any ideas?

File: cpp/src/ray/runtime/task/task_executor.cc [hunk 1]
```diff
@@ -148,18 +148,9 @@ Status TaskExecutor::ExecuteTask(
   std::shared_ptr<msgpack::sbuffer> data = nullptr;
   ArgsBufferList ray_args_buffer;
   for (size_t i = 0; i < args_buffer.size(); i++) {
-    auto &ref = arg_refs.at(i);
```
Comment (SongGuyang) [inline cpp/src/ray/runtime/task/task_executor.cc:151]: Can we also remove `arg_refs ` from args of this function? Seems it is unused after your change.

**Review by SongGuyang — COMMENTED** [review]
@jon-chuang Thank you for this fix! But I found the cpp worker CI test has failed, can you take a look? @qicosmos Can you help to review this change?

File: cpp/src/ray/runtime/task/task_executor.cc (hunk 1 above)
Comment (jon-chuang) [reply cpp/src/ray/runtime/task/task_executor.cc:151]: This function is callback with the given signature required by CoreWorker, other language workers use the data for their own reasons.

**Comment by jon-chuang** [comment]
> @jon-chuang Thank you for this fix! But I found the cpp worker CI test has failed, can you take a look? @qicosmos Can you help to review this change?

Yes, there is a SIGSEGV. So this was quite unexpected hence requiring some advice and help to diagnose if possible. As shown in the logs, the local raylet consistently dies when the task `ReturnLargeArray` is called.

File: cpp/src/ray/runtime/task/task_executor.cc (hunk 1 above)
Comment (SongGuyang) [reply cpp/src/ray/runtime/task/task_executor.cc:151]: Got.

**Comment by SongGuyang** [comment]
@qicosmos will take a look today.

**Comment by jon-chuang** [comment]
I dug into the raylet.out logs, this is the relevant bits:
```
[2021-12-22 15:26:56,519 D 206648 206648] scheduling_policy.cc:113: Node -3665707889025256223 is available for request  {(0) (0) (0) (0) }  []
 with critical resource utilization 0 based on local view  {
	CPU: (20000:20000)
	MEM: (108429926400000:108429926400000)
	GPU: (10000:10000)
	OBJECT_STORE_MEM: (54214963200000:54214963200000)
	-1:(10000:10000)
	-1:(20000:20000)
	-1:(10000:10000)
	-1:(10000:10000)
}

[2021-12-22 15:26:56,519 D 206648 206648] cluster_resource_scheduler.cc:262: Scheduling decision. forcing spillback: 0. Best node: -3665707889025256223 556eadcfba21b5d8de65601ea9a24dd62f03c0b79a7e242f333a796f, is infeasible: 0
[2021-12-22 15:26:56,519 D 206648 206648] dependency_manager.cc:171: Adding dependencies for task 4300020fbea8c68bf1790e3849fef4e32eaf1c6a01000000. Required objects length: 1
[2021-12-22 15:26:56,519 D 206648 206648] dependency_manager.cc:183: Task 4300020fbea8c68bf1790e3849fef4e32eaf1c6a01000000 blocked on object ffffffffffffffffffffffffffffffffffffffff0100000004000000
[2021-12-22 15:26:56,519 D 206648 206648] pull_manager.cc:71: Start pull request 2. Bundle size: 1
[2021-12-22 15:26:56,519 D 206648 206648] pull_manager.cc:78: Pull of object ffffffffffffffffffffffffffffffffffffffff0100000004000000
[2021-12-22 15:26:56,540 C 206648 206648] core_worker_client_pool.cc:32:  Check failed: addr_proto.worker_id() != "" 
*** StackTrace Information ***
    ray::SpdLogMessage::Flush()
    ray::RayLog::~RayLog()
    ray::rpc::CoreWorkerClientPool::GetOrConnect()
    std::_Function_handler<>::_M_invoke()
    ray::pubsub::Subscriber::SendCommandBatchIfPossible()
    ray::pubsub::Subscriber::SubscribeInternal()
    ray::pubsub::Subscriber::Subscribe()
    ray::OwnershipBasedObjectDirectory::SubscribeObjectLocations()
    ray::ObjectManager::Pull()
    ray::raylet::DependencyManager::RequestTaskDependencies()
    ray::raylet::ClusterTaskManager::WaitForTaskArgsRequests()
    ray::raylet::ClusterTaskManager::SchedulePendingTasks()
    ray::raylet::ClusterTaskManager::ScheduleAndDispatchTasks()
    ray::raylet::ClusterTaskManager::QueueAndScheduleTask()
    ray::raylet::NodeManager::HandleRequestWorkerLease()
    std::_Function_handler<>::_M_invoke()
    boost::asio::detail::completion_handler<>::do_complete()
    boost::asio::detail::scheduler::do_run_one()
    boost::asio::detail::scheduler::run()
    boost::asio::io_context::run()
    main
    __libc_start_main

```

File: cpp/src/ray/runtime/task/task_executor.cc (hunk 1 above)
Comment (qicosmos) [reply cpp/src/ray/runtime/task/task_executor.cc:151]: We need arg_refs to get the object id of arg_ref, some arguments are object ref args not a value arg.

**Comment by qicosmos** [comment]
I will run your test case to see what happened.

**Comment by jon-chuang** [comment]
Here is the sequence of events:
```
[2021-12-22 15:26:56,518 D 206440 206721] direct_task_transport.cc:532: Requesting lease from raylet 556eadcfba21b5d8de65601ea9a24dd62f03c0b79a7e242f333a796f for task 4300020fbea8c68bf1790e3849fef4e32eaf1c6a01000000
[2021-12-22 15:26:56,518 D 206440 206721] reference_count.cc:1109: Adding location 556eadcfba21b5d8de65601ea9a24dd62f03c0b79a7e242f333a796f for object ffffffffffffffffffffffffffffffffffffffff0100000004000000
---
[2021-12-22 15:26:56,519 D 206648 206648] cluster_resource_scheduler.cc:262: Scheduling decision. forcing spillback: 0. Best node: -3665707889025256223 556eadcfba21b5d8de65601ea9a24dd62f03c0b79a7e242f333a796f, is infeasible: 0
[2021-12-22 15:26:56,519 D 206648 206648] dependency_manager.cc:171: Adding dependencies for task 4300020fbea8c68bf1790e3849fef4e32eaf1c6a01000000. Required objects length: 1
[2021-12-22 15:26:56,519 D 206648 206648] dependency_manager.cc:183: Task 4300020fbea8c68bf1790e3849fef4e32eaf1c6a01000000 blocked on object ffffffffffffffffffffffffffffffffffffffff0100000004000000
[2021-12-22 15:26:56,519 D 206648 206648] pull_manager.cc:71: Start pull request 2. Bundle size: 1
[2021-12-22 15:26:56,519 D 206648 206648] pull_manager.cc:78: Pull of object ffffffffffffffffffffffffffffffffffffffff0100000004000000
[2021-12-22 15:26:56,540 C 206648 206648] core_worker_client_pool.cc:32:  Check failed: addr_proto.worker_id() != "" 
---
[2021-12-22 15:26:56,542 W 206440 206721] direct_task_transport.cc:611: The worker failed to receive a response from the local raylet because the raylet is unavailable (crashed). Error: GrpcUnavailable: RPC Error message: Socket closed; RPC Error details: 
[2021-12-22 15:26:56,542 D 206440 206721] task_manager.cc:388: Task 65a1a3aaa614cba3ffffffffffffffffffffffff01000000 failed with error LOCAL_RAYLET_DIED
```

I find it suspicious that no logs occur for 21 ms between 15:26:56,519 and 15:26:56,540

**Comment by jon-chuang** [comment]
I tried running the test but removing all of the other tasks/put/get to reduce the noise.
Here are driver logs:
```
[2021-12-23 10:17:19,941 D 341632 341632] reference_count.cc:173: Adding owned object ffffffffffffffffffffffffffffffffffffffff0100000002000000
[2021-12-23 10:17:19,941 D 341632 341632] reference_count.cc:1109: Adding location fdd21074853333ea008fb15eec86a48905ec453b4d8e074149458087 for object ffffffffffffffffffffffffffffffffffffffff0100000002000000
[2021-12-23 10:17:19,941 D 341632 341632] reference_count.cc:1307: Published message for ffffffffffffffffffffffffffffffffffffffff0100000002000000, 1 locations, spilled url: [], spilled node ID: NIL_ID, and object size: 107314, and primary node ID: fdd21074853333ea008fb15eec86a48905ec453b4d8e074149458087, pending creation? 0
[2021-12-23 10:17:19,942 D 341632 341632] client.cc:343: called plasma_create on conn 45 with size 107314 and metadata size 0
[2021-12-23 10:17:19,942 D 341632 341632] core_worker.cc:833: Pinning put object ffffffffffffffffffffffffffffffffffffffff0100000002000000
[2021-12-23 10:17:19,942 D 341632 341632] reference_count.cc:238: Add local reference ffffffffffffffffffffffffffffffffffffffff0100000002000000
[2021-12-23 10:17:19,942 D 341632 341632] reference_count.cc:239: REF ffffffffffffffffffffffffffffffffffffffff0100000002000000 borrowers: 0 local_ref_count: 1 submitted_count: 0 contained_in_owned: 0 contained_in_borrowed: 0 contains: 0 stored_in: 0 lineage_ref_count: 0
[2021-12-23 10:17:19,942 D 341632 341632] core_worker.cc:1526: Submitting normal task Type=NORMAL_TASK, Language=CPP, Resources: {}, function_descriptor={type=CppFunctionDescriptor, function_name=ReturnLargeArray}, task_id=a67dc375e60ddd1affffffffffffffffffffffff01000000, task_name=ReturnLargeArray, job_id=01000000, num_args=1, num_returns=1, depth=1
[2021-12-23 10:17:19,942 D 341632 341632] task_manager.cc:34: Adding pending task a67dc375e60ddd1affffffffffffffffffffffff01000000 with 1 retries
[2021-12-23 10:17:19,942 D 341632 341632] task_manager.cc:42: Adding arg ID ffffffffffffffffffffffffffffffffffffffff0100000002000000
[2021-12-23 10:17:19,942 D 341632 341632] reference_count.cc:173: Adding owned object a67dc375e60ddd1affffffffffffffffffffffff0100000001000000
[2021-12-23 10:17:19,942 D 341632 341632] reference_count.cc:309: Increment ref count for submitted task argument ffffffffffffffffffffffffffffffffffffffff0100000002000000
[2021-12-23 10:17:19,942 D 341632 341632] reference_count.cc:238: Add local reference a67dc375e60ddd1affffffffffffffffffffffff0100000001000000
[2021-12-23 10:17:19,942 D 341632 341632] reference_count.cc:239: REF a67dc375e60ddd1affffffffffffffffffffffff0100000001000000 borrowers: 0 local_ref_count: 1 submitted_count: 0 contained_in_owned: 0 contained_in_borrowed: 0 contains: 0 stored_in: 0 lineage_ref_count: 0
[2021-12-23 10:17:19,942 D 341632 341632] core_worker.cc:1017: ffffffffffffffffffffffffffffffffffffffff0100000002000000 in plasma, doing fetch-and-get
[2021-12-23 10:17:19,942 D 341632 341632] core_worker.cc:1032: Plasma GET timeout -1
[2021-12-23 10:17:19,942 D 341632 342062] direct_task_transport.cc:23: Submit task a67dc375e60ddd1affffffffffffffffffffffff01000000
[2021-12-23 10:17:19,942 D 341632 342062] direct_task_transport.cc:33: Task dependencies resolved a67dc375e60ddd1affffffffffffffffffffffff01000000
[2021-12-23 10:17:19,942 D 341632 342062] direct_task_transport.cc:532: Requesting lease from raylet fdd21074853333ea008fb15eec86a48905ec453b4d8e074149458087 for task 90b47c5f7d2e4d773db05efb1421fbd81c1916aa01000000
[2021-12-23 10:17:19,943 D 341632 342062] reference_count.cc:1109: Adding location fdd21074853333ea008fb15eec86a48905ec453b4d8e074149458087 for object ffffffffffffffffffffffffffffffffffffffff0100000002000000
[2021-12-23 10:17:19,943 D 341632 342062] core_worker.cc:2592: Got a long polling request from a node fdd21074853333ea008fb15eec86a48905ec453b4d8e074149458087
[2021-12-23 10:17:19,943 D 341632 342062] publisher.cc:220: Long polling connection initiated by fdd21074853333ea008fb15eec86a48905ec453b4d8e074149458087
[2021-12-23 10:17:19,970 I 341632 342062] direct_task_transport.cc:616: The worker failed to receive a response from the local raylet, but raylet is still alive. Try again on a local node. Error: IOError: 
[2021-12-23 10:17:19,970 D 341632 342062] direct_task_transport.cc:532: Requesting lease from raylet fdd21074853333ea008fb15eec86a48905ec453b4d8e074149458087 for task 4e20116f1f71d71e9244738dbbc4e98d459c999501000000
```
here are raylet logs:
```
[2021-12-23 10:17:19,943 D 341987 342011] store.cc:250: Object ffffffffffffffffffffffffffffffffffffffff0100000002000000 no longer in use by client
[2021-12-23 10:17:19,944 I 341987 341987] worker_pool.cc:414: Started worker process of 1 worker(s) with pid 342068
[2021-12-23 10:17:19,944 D 341987 341987] cluster_task_manager.cc:481: Queuing and scheduling task 90b47c5f7d2e4d773db05efb1421fbd81c1916aa01000000
[2021-12-23 10:17:19,944 D 341987 341987] cluster_task_manager.cc:80: Scheduling pending task 90b47c5f7d2e4d773db05efb1421fbd81c1916aa01000000
[2021-12-23 10:17:19,944 D 341987 341987] scheduling_policy.cc:112: Node -5751289363294905732 is available for request  {(0) (0) (0) (0) }  []
 with critical resource utilization 0 based on local view  {
	CPU: (20000:20000)
	MEM: (105985302540000:105985302540000)
	GPU: (10000:10000)
	OBJECT_STORE_MEM: (52992651260000:52992651260000)
	-1:(20000:20000)
	-1:(10000:10000)
	-1:(10000:10000)
	-1:(10000:10000)
}

[2021-12-23 10:17:19,944 D 341987 341987] cluster_resource_scheduler.cc:256: Scheduling decision. forcing spillback: 0. Best node: -5751289363294905732 fdd21074853333ea008fb15eec86a48905ec453b4d8e074149458087, is infeasible: 0
[2021-12-23 10:17:19,944 D 341987 341987] dependency_manager.cc:171: Adding dependencies for task 90b47c5f7d2e4d773db05efb1421fbd81c1916aa01000000. Required objects length: 1
[2021-12-23 10:17:19,944 D 341987 341987] dependency_manager.cc:183: Task 90b47c5f7d2e4d773db05efb1421fbd81c1916aa01000000 blocked on object ffffffffffffffffffffffffffffffffffffffff0100000002000000
[2021-12-23 10:17:19,944 D 341987 341987] pull_manager.cc:70: Start pull request 1. Bundle size: 1
[2021-12-23 10:17:19,944 D 341987 341987] pull_manager.cc:77: Pull of object ffffffffffffffffffffffffffffffffffffffff0100000002000000
[2021-12-23 10:17:19,968 C 341987 341987] core_worker_client_pool.cc:32:  Check failed: addr_proto.worker_id() != "" 
```
Again, there is a long pause, and a lot of missing information about what happens in between.

**Comment by qicosmos** [comment]
Thanks for your testing.
If you change the test case like this with master c++ worker:
```
  std::array<int, 100000> arr;
  auto r18 = ray::Task(ReturnLargeArray).Remote(arr);
  EXPECT_EQ(arr, *(ray::Get(r18)));
```
the test case will pass, i think it's related with object ref arguments, i will check it later.

**Comment by jon-chuang** [comment]
I'm not sure this tests the same code pathway, as the object might not first enter the object store?

**Comment by jon-chuang** [comment]
Here is my suspicion for the fix: when passing objectref to remote, increase the reference counter...?

**Comment by jon-chuang** [comment]
The behaviour I seem to be seeing is that something is triggering `ReleaseObject` prior to that object ref being resolved for the downstream task.

**Comment by jon-chuang** [comment]
Indeed, it seems to me that the C++ implementation is not up to spec; when passing an object by reference to a task, one ought to increment a pending task count (not a local ref count).

**Comment by jon-chuang** [comment]
I think this is the error:
```
ray_arg = absl::make_unique<ray::TaskArgByReference>(ObjectID::FromBinary(*arg.id),
                                                           ray::rpc::Address{}, // this should not be empty
                                                           /*call_site=*/"");
```
I think there are a lot of errors in C++ API implementation. More thorough testing should be done to catch such errors.

I also think there is a lot of confusion in the language implementation as reflected by code structure and lack of comments. I believe that a language implementation spec/guide/checklist, and more thoroughness in commenting and ensuring that the calling contract conditions are met can help reduce these errors.

**Comment by jon-chuang** [comment]
So there was another bug in the C++ API implementation, which was not adding the object owner address to `TaskArgByReference` when calling `Remote`.

**Comment by qicosmos** [comment]
Hi, i'm also solving the reference counting bug before
https://github.com/ray-project/ray/pull/20143
could you also test the use case:
https://github.com/ray-project/ray/blob/878dd15fb09284967974e90d9a95662151060c90/cpp/src/ray/test/cluster/cluster_mode_test.cc#L407

```
  auto obj = ray::Put(std::string("aaa"));
  auto r = ray::Task(GetVal).Remote(obj);
  EXPECT_EQ(*r.Get(), "aaa");

  auto obj1 = ray::Put(1);
  auto obj2 = ray::Put(2);
  auto r1 = ray::Task(Add).Remote(obj1, obj2);
  EXPECT_EQ(*r1.Get(), 3);

  std::vector v{obj1, obj2};
  auto r2 = ray::Task(GetList).Remote(v);
  EXPECT_EQ(*r2.Get(), 3);```

**Comment by jon-chuang** [comment]
Why not we merge this PR first, and in the future, change to use the cached `owner_address_` for Native mode?

**Comment by qicosmos** [comment]
> Why not we merge this PR first, and in the future, change to using the `ObjectRef` cached `owner_address_` (only for Native mode) instead of dynamically retrieving from `CoreWorker` as done here?

Thanks for your fix.

**Review by qicosmos — APPROVED** [review]
LGTM

File: cpp/src/ray/runtime/abstract_ray_runtime.cc [hunk 2]
```diff
@@ -129,8 +129,13 @@ std::vector<std::unique_ptr<::ray::TaskArg>> TransformArgs(
           memory_buffer, nullptr, std::vector<rpc::ObjectReference>()));
     } else {
       RAY_CHECK(arg.id);
-      ray_arg = absl::make_unique<ray::TaskArgByReference>(ObjectID::FromBinary(*arg.id),
-                                                           ray::rpc::Address{},
+      auto id = ObjectID::FromBinary(*arg.id);
+      auto owner_address = ray::rpc::Address{};
+      if (CoreWorkerProcess::IsInitialized()) {
```
Comment (kfstorm) [inline cpp/src/ray/runtime/abstract_ray_runtime.cc:134]: Is there any reason to check this?

Comment (jon-chuang) [reply cpp/src/ray/runtime/abstract_ray_runtime.cc:134]: I think it is workaround for local mode. Maybe it ought to check if the config is native mode instead?

Comment (kfstorm) [reply cpp/src/ray/runtime/abstract_ray_runtime.cc:134]: That makes sense. Could you use `ConfigInternal::Instance().run_mode == RunMode::CLUSTER` instead?

**Comment by jon-chuang** [comment]
CI failures seem build/package related and unrelated to this PR...
e.g.
```
ERROR: Could not find a version that satisfies the requirement pettingzoo==1.14.0 (from versions: 0.1.0, 0.1.1, 0.1.2, 0.1.3, 0.1.4, 0.1.5, 0.1.6, 0.1.7, 0.1.8, 0.1.9, 0.1.10, 0.1.11, 0.1.12, 0.1.13, 0.1.14, 1.0.0, 1.0.1, 1.1.0, 1.2.0, 1.2.1, 1.3.0, 1.3.1, 1.3.2, 1.3.3, 1.3.4, 1.3.5, 1.4.0, 1.4.1, 1.4.2, 1.5.0, 1.5.1, 1.5.2, 1.6.0, 1.6.1, 1.7.0, 1.8.0, 1.8.1, 1.8.2, 1.9.0, 1.10.0, 1.11.0, 1.11.1, 1.11.2, 1.12.0)
--
  | ERROR: No matching distribution found for pettingzoo==1.14.0
```
Not sure how to fix

**Comment by SongGuyang** [comment]
@jon-chuang Rebase the master and make the CI test run again? And you can check the flakey test in https://flakey-tests.ray.io/.

**Comment by jon-chuang** [comment]
Do you want me to rebase or fine to just merge master?

**Comment by jon-chuang** [comment]
bump to merge this, please advise what to do in spite of flaky tests clearly not related to this PR, and whether to merge or rebase master, or no preference. Thank you.

**Comment by kfstorm** [comment]
> Do you want me to rebase or fine to just merge master?
> 
> Not sure what is the preferred practice.

I believe merge is preferred. Rebase will change commit history and it makes reviewers hard to know what's changed since the last review.

Sorry for the late reply. Merging.

**Pull request merged**